src/anyp/Uri.cc

```diff
@@ -81,6 +103,50 @@ AnyP::Uri::Encode(const SBuf &buf, const CharacterSet &ignore)
     return output;
 }
 
+SBuf
+AnyP::Uri::Rfc3986Encode(const SBuf &buf)
+{
+    return Encode(buf, CharacterSet::RFC3986_UNRESERVED);
+}
+
+SBuf
+AnyP::Uri::Decode(const SBuf &buf)
+{
+    if (buf.isEmpty())
+        return buf;
+
+    SBuf output;
+    output.reserveSpace(buf.length()); // worst case: buf is not encoded
+
+    Parser::Tokenizer tok(buf);
+    SBuf token;
+    static const auto unencodedOctets = CharacterSet("unencoded", "%").complement();
+
+    tok.prefix(token, unencodedOctets);
+    output.append(token);
+
+    while (!tok.atEnd())
+    {
+        if (!tok.skip('%')) // should never happen
+            throw TextException("incomplete %-encoded triplet at end of input", Here());
```

Comment:
I'm trying to navigate between your and Alex' suggestion; while I welcome the scrutiny over my proposal, let's walk through the code because I think that while the message could very well be improved, the code itself should be okay

1.  Lines 125-126's purpose is ensure that we are at the end of input because there's no '%' in the input, or that the first character of the unparsed input is a '%'.
2. if it's the former case, then the condition at line 128 is false, the loop is never executed, and we just return the input.
3. If it's the latter, then the conditions at line 130 should always be true. We could argue that this could be an assertion instead an exception, or that the text could be clearer. I'm happy to follow any suggestion here
4. lines 133-134 extract the encoded hex characters, and there's exactly two. The decoding is fairly straightforward
5. the invariant [2] should be guaranteed at every iteration of the while loop, by the conditions at lines 144-145. The test at line 144 is meant to guard against the case where the encoded digits are at the end of input; in that case the token output parameter would be let unchanged resulting in duplicate output

The main difference with your proposal is that the invariant at [2] simplifies the code structure of the inner loop, avoiding the need of multiple atEnd() tests

Reply:
> Francesco: I'm trying to navigate between your and Alex' suggestion

I assume you are referring to https://github.com/squid-cache/squid/pull/1626#discussion_r1437200159 as my suggestion. FWIW, I appreciate you removing atEnd() checks inside the loop, but there are still too many `if` statements AFAICT. I did not check further. To minimize conflicting suggestions, I will come back to this PR after you make Amos happy.

> We could argue that this could be an assertion instead an exception

Rules of thumb:

* If X should never happen regardless of input, then Assure(!X) or assert(!X). In most input parsing cases, Assure() is the right choice because the problem is likely confined to the problematic transaction and the nearby code already throws for other reasons.
* If X may only happen when parsing invalid input, then throw TextException.